src/CollectWidget.res

```diff
@@ -51,30 +92,78 @@ let make = (
       | Some(Card(_)) => {
           setSelectedPaymentMethod(_ => None)
           setSelectedPaymentMethodType(_ => None)
-          resetForm()
         }
       | Some(_) => setSelectedPaymentMethodType(_ => None)
       | None =>
         switch selectedPaymentMethod {
-        | Some(_) => {
-            setSelectedPaymentMethod(_ => None)
-            resetForm()
-            ()
-          }
+        | Some(_) => setSelectedPaymentMethod(_ => None)
         | None => ()
         }
       }
     }
   }
 
-  let setPaymentMethodDataValue = (key: paymentMethodDataField, value) =>
-    setPaymentMethodData(_ => paymentMethodData->setValue(key->getPaymentMethodDataFieldKey, value))
-
   let getPaymentMethodDataValue = (key: paymentMethodDataField) =>
     paymentMethodData
     ->getValue(key->getPaymentMethodDataFieldKey)
     ->Option.getOr("")
 
+  let setPaymentMethodDataValue = (key: paymentMethodDataField, value) =>
+    setPaymentMethodData(_ => paymentMethodData->setValue(key->getPaymentMethodDataFieldKey, value))
+
+  let validateAndSetPaymentMethodDataValue = (key: paymentMethodDataField, event) => {
+    let value = ReactEvent.Form.target(event)["value"]
+    let inputType = ReactEvent.Form.target(event)["type"]
+
+    let (isValid, updatedValue) = switch (key, inputType, value) {
+    // Empty string is valid (no error)
+    | (_, _, "") => (true, "")
+    | (CardExpDate, "number" | "tel", _) => {
+        let formattedExpiry = formatCardExpiryNumber(value)
+        if isExipryValid(formattedExpiry) {
+          handleInputFocus(~currentRef=cardExpRef, ~destinationRef=cardHolderRef)
+        }
+        (true, formattedExpiry)
+      }
+    | (CardNumber, "number" | "tel", _) => {
+        let cardType = getCardType(getPaymentMethodDataValue(CardBrand))
+        let formattedCardNumber = formatCardNumber(value, cardType)
+        if cardValid(clearSpaces(formattedCardNumber), getCardStringFromType(cardType)) {
+          handleInputFocus(~currentRef=cardNumberRef, ~destinationRef=cardExpRef)
+        }
+        (true, formattedCardNumber)
+      }
+    | (SepaBic | SepaIban, "text", _) => (true, String.toUpperCase(value))
+
+    // Default number validation
+    | (_, "number" | "tel", _) =>
+      try {
+        let bigIntValue = Js.BigInt.fromStringExn(value)
+        (true, Js.BigInt.toString(bigIntValue))
+      } catch {
+      | _ => (false, value)
+      }
+
+    // Default validation
+    | (_, _, _) =>
+      getPaymentMethodDataFieldCharacterPattern(key)
+      // valid; in case there is no pattern setup
+      ->Option.mapOr((true, value), regex =>
+        regex->RegExp.test(value) ? (true, value) : (false, value)
+      )
+    }
+
+    if isValid {
+      switch key {
+      | CardNumber => {
+          setPaymentMethodDataValue(CardBrand, getCardBrand(updatedValue))
+          setPaymentMethodDataValue(key, updatedValue)
```

Comment:
wait, why are you calling `setPaymentMethodDataValue()` twice in case of CardNumber

Reply:
I'm setting two fields here - `CardBrand` and `CardNumber`